### DISCUSSION TITLE
Query | Reusable NgEntityService

### DISCUSSION THREAD

**Discussion by vikasher** [salesforce/akita#715] [Q&A] [answered]
## I'm submitting a question

I'm trying to create a reusable NgEntityService, My use case is I have to manage three entities `ActiveUsers`, `PurgedUsers`, `InactiveUsers`, All three consumes the same API and have same base url only queryparam varies while fetching.
Unfortunately I'm stuck , I tried looking on SO but couldn’t find anything.
Is there any way this can be achieved?

This is what I have tried so far 

```
/**
 * Entity Service
 */
type Store = ActiveUsersStore | PurgedUsersStore | InactiveUsersStore;
type State = ActiveUsersState | PurgedUsersState | InactiveUsersState;

@NgEntityServiceConfig({ resourceName: 'api/users' })
export class UsersService extends NgEntityService<State> {
  constructor(protected store: Store, configService: ConfigService) {
    super(store);
    this.baseUrl = configService.getBaseUrl();
  }
}

const activeUserfactory = (
  store: ActiveUsersStore,
  configService: ConfigService
) => {
  return new UsersService(store, configService);
};

@Component({
  selector: 'my-app',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
  providers: [
    {
      provide: UsersService,
      useFactory: activeUserfactory,
      deps: [ActiveUsersStore, ConfigService]
    }
  ]
})
export class ActiveComponent {
  constructor(public userService: UsersService) {}
}

```
[*StackBlitz*](https://stackblitz.com/edit/akita-entity-service-gt6bxr)

Thanks in advance.

**Comment by vikasher** [comment]
@NetanelBasal  Could you please help me with this , I do follow you on medium and twitter and I have gone through all your blogs.
Excerpt from [Angular Reactive Forms: Tips and Tricks](https://netbasal.com/angular-reactive-forms-tips-and-tricks-bb0c85400b58). `I also wanted to mention that I’m against premature optimization— you should always prefer readability and reusability, and only apply optimizations when you think you may need them.` 
 I guess this does apply here as creating entity service per each entity wouldn’t hurt the performance . Nevertheless I was wondering if there exist a way I can avoid code duplication.? assume  in case of 10+ entities? 

> ## I'm submitting a question
> I'm trying to create a reusable NgEntityService, My use case is I have to manage three entities `ActiveUsers`, `PurgedUsers`, `InactiveUsers`, All three consumes the same API and have same base url only queryparam varies while fetching.
> Unfortunately I'm stuck , I tried looking on SO but couldn’t find anything.
> Is there any way this can be achieved?
> 
> This is what I have tried so far
> 
> ```
> /**
>  * Entity Service
>  */
> type Store = ActiveUsersStore | PurgedUsersStore | InactiveUsersStore;
> type State = ActiveUsersState | PurgedUsersState | InactiveUsersState;
> 
> @NgEntityServiceConfig({ resourceName: 'api/users' })
> export class UsersService extends NgEntityService<State> {
>   constructor(protected store: Store, configService: ConfigService) {
>     super(store);
>     this.baseUrl = configService.getBaseUrl();
>   }
> }
> 
> const activeUserfactory = (
>   store: ActiveUsersStore,
>   configService: ConfigService
> ) => {
>   return new UsersService(store, configService);
> };
> 
> @Component({
>   selector: 'my-app',
>   templateUrl: './app.component.html',
>   styleUrls: ['./app.component.css'],
>   providers: [
>     {
>       provide: UsersService,
>       useFactory: activeUserfactory,
>       deps: [ActiveUsersStore, ConfigService]
>     }
>   ]
> })
> export class ActiveComponent {
>   constructor(public userService: UsersService) {}
> }
> ```
> 
> [_StackBlitz_](https://stackblitz.com/edit/akita-entity-service-gt6bxr)
> 
> Thanks in advance.

@NetanelBasal  Could you please help me with this , I do follow you on medium and twitter and I have gone through all your blogs.
Excerpt from [Angular Reactive Forms: Tips and Tricks](https://netbasal.com/angular-reactive-forms-tips-and-tricks-bb0c85400b58). `I also wanted to mention that I’m against premature optimization— you should always prefer readability and reusability, and only apply optimizations when you think you may need them.` 
 I guess this does apply here as creating entity service per each entity wouldn’t hurt the performance . Nevertheless I was wondering if there exist a way I can avoid code duplication.? assume  in case of 10+ entities?

**Comment by NetanelBasal** [comment]
You can create a super class that extends NgEntityService and share it.

**Reply by vikasher** [reply]
One more small query, Is is possible to set the base url per module basis. As of now I'm calling `this.setBaseUrl` method in entity service. When I try provide the config in child module , I get the error , Can't find base URL?
```
    {
      provide: NG_ENTITY_SERVICE_CONFIG,
      useValue: { baseUrl: '' },// set in child module
    }

**Comment by vikasher** [comment, marked as answer]
@NetanelBasal  Thanks for your help , I was able to solve it by creating a base class.
